### PR TITLE
#293 make duplicate public key id invalid

### PR DESCRIPTION
**Pull request by isaacJChen**

### CODE REVIEW

**Comment by isaacJChen** [comment]
test

File: tests/core/Observer.spec.ts [hunk 1]
```diff
@@ -143,8 +143,8 @@ describe('Observer', async () => {
     // Prepare the mock response from the DownloadManager.
     const didDocumentTemplate = require('../json/didDocumentTemplate.json');
 
-    const [publicKey1, privateKey1] = await Cryptography.generateKeyPairHex('#key1', KeyUsage.recovery);
-    const [publicKey2, privateKey2] = await Cryptography.generateKeyPairHex('#key2', KeyUsage.recovery);
+    const [publicKey1, privateKey1] = await Cryptography.generateKeyPairHex('#extraRecoveryKey1', KeyUsage.recovery);
+    const [publicKey2, privateKey2] = await Cryptography.generateKeyPairHex('#extraRecoveryKey2', KeyUsage.recovery);
```
Comment (thehenrytsai) [inline tests/core/Observer.spec.ts]: As discussed offline, lets get rid of `generateCreateOperation()` and used the declaritive approach of generating document. ie. `Document.create()`

File: tests/core/Operation.spec.ts [hunk 2]
```diff
@@ -4,6 +4,7 @@ import KeyUsage from '../../lib/core/versions/latest/KeyUsage';
 import Operation from '../../lib/core/versions/latest/Operation';
 import OperationGenerator from '../generators/OperationGenerator';
 import { SidetreeError } from '../../lib/core/Error';
+import DocumentModel from '../../lib/core/versions/latest/models/DocumentModel';
```
Comment (thehenrytsai) [inline tests/core/Operation.spec.ts]: Nit: we normally sort imports.

File: tests/core/MongoDbOperationStore.spec.ts [hunk 3]
```diff
@@ -36,8 +38,23 @@ async function constructAnchoredCreateOperation (
...
+      'type': 'IdentityHub',
+      'serviceEndpoint': {
+        '@context': 'schema.identity.foundation/hub',
+        '@type': 'UserServiceEndpoint',
+        'instance': [
+          'did:sidetree:value0'
+        ]
+      }
+    }
+  ];
```
Comment (thehenrytsai) [inline tests/core/MongoDbOperationStore.spec.ts]: Can we create a method `OperationGenerator.createIdentityHubServiceEndpoint(instances: string[])`?

Then:

```javascript
const identityHubServiceEndpoint =  OperationGenerator.createIdentityHubServiceEndpoint(['did:sidetree:value0']);
const services = [identityServiceEndpoint];
```

File: tests/core/Observer.spec.ts [hunk 4]
```diff
@@ -141,15 +143,16 @@ describe('Observer', async () => {
...
+    const signingKeys = await Cryptography.generateKeyPairHex('#key2', KeyUsage.signing);
+    const publicKeys = [
+      publicKey1,
+      signingKeys[0]
     ];
-    const operationsBuffer = operations.map((op) => { return Buffer.from(JSON.stringify(op)); });
+    const service = OperationGenerator.createIdentityHubServiceEndpoints(['did:sidetree:value0']);
+    const document = Document.create(publicKeys, service);
+    const operationsBuffer = [await OperationGenerator.createOperationBuffer(OperationType.Create, document, '#key1', privateKey1)];
+
```
Comment (mudiali) [inline tests/core/Observer.spec.ts]: As per our offline conversation, maybe you can create the following function in OperationGenerator.ts which will be more explicit than the previous one and still be reusable. 

````
async createOperationBuffer(
  recoveryPublicKey: DidPublicKeyModel, 
  recoveryPrivateKey: string, 
  signingPublicKey: DidPublicKeyModel, 
  signingPrivateKey: string, 
  serviceEndpoints: string[]) : Promise<Buffer>{
...
}

File: tests/core/Observer.spec.ts [hunk 5]
```diff
@@ -141,15 +143,16 @@ describe('Observer', async () => {
...
-      await OperationGenerator.generateCreateOperation(didDocumentTemplate, publicKey2, privateKey2)
+    const signingKeys = await Cryptography.generateKeyPairHex('#key2', KeyUsage.signing);
+    const publicKeys = [
+      publicKey1,
+      signingKeys[0]
     ];
-    const operationsBuffer = operations.map((op) => { return Buffer.from(JSON.stringify(op)); });
+    const service = OperationGenerator.createIdentityHubServiceEndpoints(['did:sidetree:value0']);
+    const document = Document.create(publicKeys, service);
+    const operationsBuffer = [await OperationGenerator.createOperationBuffer(OperationType.Create, document, '#key1', privateKey1)];
```
Comment (mudiali) [inline tests/core/Observer.spec.ts]: Previously, the test was creating 2 operation buffers. Let's keep that logic and create 2 buffers.

File: tests/generators/OperationGenerator.ts [hunk 6]
```diff
@@ -242,4 +216,22 @@ export default class OperationGenerator {
...
+   */
+  public static createIdentityHubServiceEndpoints (instances: string[]): any[] {
+    return [
+      {
+        'id': 'IdentityHub',
+        'type': 'IdentityHub',
+        'serviceEndpoint': {
+          '@context': 'schema.identity.foundation/hub',
+          '@type': 'UserServiceEndpoint',
+          'instance': instances
```
Comment (mudiali) [inline tests/generators/OperationGenerator.ts]: Looks like we are only creating the 'UserServiceEndpoint' (see ````'@type': 'UserServiceEndpoint'````), can we change the function name to reflect that please? 

Also, seems like that the 'instance' property is supposed to be single-valued right? If single valued, then let's change the input to ````instance: string````

Comment (isaacJChen) [reply tests/generators/OperationGenerator.ts]: I will double check with the rest of the team to make sure if it's supposed to be singular or plural.

Comment (isaacJChen) [reply tests/generators/OperationGenerator.ts]: I changed it to instances for now and everything seems to be working fine.

Comment (isaacJChen) [reply tests/generators/OperationGenerator.ts]: Changed instances and related validations.

**Pull request merged**